src/modules/adhs/src/lib/Adhs.c

```diff
@@ -0,0 +1,360 @@
+        {
+            OsConfigLogError(AdhsGetLog(), "MmiGet(%s, %s) insufficient maxmimum size (%d bytes) versus data size (%d bytes), reported value will be truncated", componentName, objectName, g_maxPayloadSizeBytes, *payloadSizeBytes);
+            *payloadSizeBytes = g_maxPayloadSizeBytes;
+        }
+
+        *payload = (MMI_JSON_STRING)malloc(*payloadSizeBytes + 1);
+        if (NULL != *payload)
+        {
+            strncpy(*payload, value, *payloadSizeBytes);
+        }
+        else
+        {
+            OsConfigLogError(AdhsGetLog(), "MmiGet: failed to allocate %d bytes", *payloadSizeBytes + 1);
+            *payloadSizeBytes = 0;
+            status = ENOMEM;
+        }
+    }
+
+    if (IsFullLoggingEnabled())
+    {
+        OsConfigLogInfo(AdhsGetLog(), "MmiGet(%p, %s, %s, %.*s, %d) returning %d", clientSession, componentName, objectName, *payloadSizeBytes, *payload, *payloadSizeBytes, status);
+    }
+
+    return status;
+}
+
+int AdhsMmiSet(MMI_HANDLE clientSession, const char* componentName, const char* objectName, const MMI_JSON_STRING payload, const int payloadSizeBytes)
+{
+    int status = MMI_OK;
+
+    if ((NULL == componentName) || (NULL == objectName) || (NULL == payload) || (payloadSizeBytes < 0))
+    {
+        OsConfigLogError(AdhsGetLog(), "MmiSet(%s, %s, %p, %d) called with invalid arguments", componentName, objectName, payload, payloadSizeBytes);
+        status = EINVAL;
+        return status;
+    }
+
+    if (!IsValidSession(clientSession))
+    {
+        OsConfigLogError(AdhsGetLog(), "MmiSet(%s, %s) called outside of a valid session", componentName, objectName);
+        status = EINVAL;
+    }
+        
+    if ((MMI_OK == status) && (strcmp(componentName, g_adhsComponentName)))
+    {
+        OsConfigLogError(AdhsGetLog(), "MmiSet called for an unsupported component name (%s)", componentName);
+        status = EINVAL;
+    }
+        
+    if ((MMI_OK == status) && (strcmp(objectName, g_desiredOptInObjectName)))
+    {
+        OsConfigLogError(AdhsGetLog(), "MmiSet called for an unsupported object name (%s)", objectName);
+        status = EINVAL;
+    }
+        
+    if ((MMI_OK == status) && (!IsValidPayload(payload, payloadSizeBytes)))
+    {
+        OsConfigLogError(AdhsGetLog(), "MmiSet(%.*s, %d) called with invalid payload", payloadSizeBytes, payload, payloadSizeBytes);
+        status = EINVAL;
+    }
+
+    if (MMI_OK == status)
+    {
+        const char* value = NULL; 
+        for (unsigned int i = 0; i < g_permissionConfigMapCount; i++)
+        {
+            if ((payloadSizeBytes == (int)strlen(g_permissionConfigMapValues[i])) && (0 == strncmp(payload, g_permissionConfigMapValues[i], payloadSizeBytes)))
+            {
+                value = g_permissionConfigMapKeys[i];
+                break;
+            }
+        }
+
+        if (NULL != value)
+        {
+            const char* format = "Permission = \"%s\"\n";

```

Comment:
>const char* format = "Permission = \"%s\"\n" [](http://example.com/codeflow?start=12&length=44)

Make this a global constant. Also, for C, all local variables, including constants, should be declared at beginning of function.

Reply:
Thanks, fixed!